Improve comment readability

sauerlandf opened the pull request:
changed head branch to avoid including pull request #52 . This pull request replaces pull request #50

sauerlandf committed:
improve comment readability

sauerlandf committed:
remove inline jquery scripts

sauerlandf committed:
make override more specific

sauerlandf committed:
optimize codeflow in listAttendances.blade.php

also changes behaviour of toggling

Commit:
show number of 'maybe' in GigAttendanceList

* show number of 'maybe' in GigAttendanceLise

Including speedup: Complexity of queries decreased (approx. half execution time).

* Update app.css

dummy commit to avoid unnecessary merge conflicts upon branching

* Update .gitignore

* Update .gitignore

* remove IDE artifacts

* remove IDE artifacts

sauerlandf committed:
debug calender date view call from main menu

sauerlandf committed:
add overview for reheaersal attendances

nielsranosch commented on .buildpath:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<buildpath>
+	<buildpathentry kind="src" path=""/>
+	<buildpathentry kind="con" path="org.eclipse.php.core.LANGUAGE"/>
+</buildpath>
```
"git rm" this file

sauerlandf replied:
geez, i thought i was done doing all that cleanup

nielsranosch commented on app/Models/Event.php:
```diff
@@ -173,6 +173,48 @@ public function getAttendanceCount($users = null) {
+     * @param $attendances - a Collection of Attendances
+     * @return int[] - an array with the number of attendances. Use e.g. int[2] to get number of "yes"-Answers
+     */
+    public static function getAttendanceCountNew($attendances) {
```
That is no way to name a function.

sauerlandf replied:
true. I'm not sure if getAttendanceCount is in use somewhere else though, which is why i gave this function a temporary name. I might just figure out an entirely new name for this.

nielsranosch commented on resources/assets/sass/app.scss:
```diff
@@ -1163,6 +1185,22 @@ a.inline-link {
       background-color: $error-color;
     }
   }
+
+  /*
+  tr.subvoice td{
+    .overviewnumber{
+      opacity: 0.8;
+      padding: 0.1em;
+    }
+
+    .positive{
+      background-color: $success-color;
+    }
+
+    .maybe{
+      background-color: $warning-color;
+    }
+  }*/
```
Not sure what you are trying to do here ;-)

sauerlandf replied:
i originally added colours to the overview numbers, but it ended up looking weird. You (or I) can safely remove this

nielsranosch replied:
Yes, please :-)

nielsranosch commented on resources/views/date/gig/listAttendances.blade.php:
```diff
@@ -29,44 +29,82 @@
+                                                        //TODO: This should probably go into GigAttendanceController somehow...
+                                                        $voiceAttendances = $gigattendances[$gig->id];
+                                                        $voiceAttendances = \App\Models\Event::filterAttendancesByUserIDs($voiceAttendances, $userIDs);
+                                                        $voiceAttendances = \App\Models\Event::getAttendanceCountNew($voiceAttendances);
```
It should. It's fine for now, I would be ok to merge it.

nielsranosch commented on app/Http/Controllers/RehearsalAttendanceController.php:
```diff
@@ -53,12 +54,25 @@ protected function prepareAdditionalData(Request $request) {
      * @return $this|\Illuminate\Contracts\View\Factory|\Illuminate\View\View
      */
     public function listAttendances ($rehearsal_id = null) {
+        if ('all' === $rehearsal_id){
+            // Get all future rehearsals.
+            $rehearsals = Rehearsal::with('rehearsal_attendances.user')->where('end', '>=', Carbon::today())->orderBy('start')->paginate(8, ['title', 'start', 'id']);
+            if (null === $rehearsals || sizeof($rehearsals) < 1) {
+                return back()->withErrors(trans('date.no_rehearsals_in_future'));
+            }
+            $voices = Voice::getParentVoices();
+            return view('date.rehearsal.listAllAttendances', [
+                'rehearsals'  => $rehearsals,
+                'voices' => $voices,
+            ]);
+        }
+
```
Add //TODO: maybe merge this function with the same function in GigAttendanceController to a unified function in AttendanceController

sauerlandf replied:
done!

nielsranosch commented on app/Models/Event.php:
```diff
@@ -173,6 +173,48 @@ public function getAttendanceCount($users = null) {
+     * Filter a Collection of Attendances by an Array of Keys from Users
+     * 
+     * @param Attendance[] $attendances
+     * @param int[]|<key>[] $userIDs
+     * @return Attendance[]
+     */
+    public static function filterAttendancesByUserIDs($attendances, $userIDs){
+        if (sizeof($userIDs) < 1){
+            return null;
+        }
```
In Laravel, we try to us Collections instead of Arrays weherever possible. One of the reasons is, that sizeof() is not always reliable. Is it possible to use a Collection here (also in the other function).

sauerlandf replied:
yeah, i expected `$users->keyBy('id')->keys()->all()` to be a lot slower than it actually is. I've added this function so I don't have to call above code for every event, but the runtime difference is minicule, so I'm gonna revert this and remove this function.

nielsranosch commented on package.json:
```diff
@@ -2,13 +2,17 @@
     "laravel-elixir": "^6.0.0-18"
+  },
+  "resolutions": {
+    "graceful-fs": "4.2.3"
```
Add comment explainig why this is here (workaround for https://github.com/nielsranosch/jc_intern/issues/51 )

sauerlandf replied:
Comments are not supported in json... added this to .gitignore for now

nielsranosch commented on resources/views/date/rehearsal/listAllAttendances.blade.php:
```diff
@@ -0,0 +1,153 @@
+                                                @foreach($rehearsals as $rehearsal)
+                                                    <td>
+                                                    <?php 
+                                                        //TODO: This should probably go into GigAttendanceController somehow...
```
Same as above

nielsranosch replied:
Typo ;-)

nielsranosch commented on resources/views/date/rehearsal/listAllAttendances.blade.php:
```diff
@@ -0,0 +1,153 @@
+            $(".btn-toggle").click(function () {
+                var voice = $(this).data('voice');
+
+                //TODO: Make this more beautiful.
```
Fine for now. Feel free to improve it once you know more about jQuery ;-)

sauerlandf replied:
This is literally your code (+comment) that I just have copied ;)

Anyway, in the latest version, this code has been changed to something more beautiful

nielsranosch commented on resources/views/date/gig/listAttendances.blade.php:
```diff
@@ -29,44 +29,82 @@
+                                                                <?php $comment = $gig->getComment($user);?>
+                                                                &nbsp;
+                                                                <i class="far fa-comment comment-toggle" title="{{$comment}}"></i>
+                                                                <div class="full-comment" style="display: none"> {{$comment}} </div>
```
I don't like that this is clickable without any visual indication that it is. I will prepare a commit to show you what I mean.

sauerlandf replied:
see my comment on your commit. I like your visuals better, but i added the extra clickability for a reason, so I won't change this for now

nielsranosch reviewed (changes requested):
Always assume that a function name will be there forever. That essentially means you have two choices: Make it abstract or make it specific. In either case, your name should be descriptive enough.

I am getting an Exception for undefined index in app\Models\Event.php185 when accessing http://localhost/public/gigs/attendances/list/141.

These are all just small things. All in all, very strong start! Good ideas for how things should flow, which will just improve over time. Think about different solutions (especially in jQuery) and weigh them against each other. Some things work better than others.

nielsranosch committed:
We prefer to use <a> instead of changing the cursor (wherever possible). Also now there is a small button. What do you think?

sauerlandf committed:
in gig and rehearsal overviews, make comments and subvoice toggles larger

for mobile devices in particular, otherwise, the user would have to scroll through the table too much

sauerlandf committed:
Merge branch 'comments_readability_temp' into comments_readability

sauerlandf committed:
added comments, revert package.json changes

sauerlandf committed:
cleanup code

sauerlandf committed:
remove locally required changes to package.json

sauerlandf committed:
update package.json

sauerlandf committed:
unify views

nielsranosch commented on resources/views/date/gig/listAttendances.blade.php:
```diff
@@ -88,19 +127,22 @@
+                    togglebutton.data("status", "hidden").children("i").removeClass("fa-caret-down").addClass("fa-caret-right");
+                }
+            });
+            // open and close comments for one user
+            $("td").has(".comment-toggle").click(function(){
+                $(this).siblings().addBack().find(".full-comment").toggle();
             });
```
I don't think it's a good design choice that the comment disappears when one tries to highlight it. I think the button is big enough and the onclick-function for the comment itself should be removed.

nielsranosch commented on resources/views/date/rehearsal/listAllAttendances.blade.php:
```diff
@@ -0,0 +1,151 @@
+                                                </tr>
+                                            @endforeach
+                                        @endforeach
+                                    @endforeach
+                                </tbody>
+                            </table>
+                        </div>
+                        @if(!is_array($rehearsals))
+                            {{ $rehearsals->links() }}
+                        @endif
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+@endsection
+
+@section('js')
+    <script type="text/javascript">
+        $(document).ready(function () {
+            // open and close subvoices
+            $(".subvoice").click(function (){
+                var togglebutton = $(this).find(".btn-toggle");
+                var voice = togglebutton.data('voice');
+
+                $(this).nextAll(".voice-"+voice).toggle();
+                if("hidden" === togglebutton.data("status")){
+                    togglebutton.data("status","display").children("i").removeClass("fa-caret-right").addClass("fa-caret-down");
+                }
+                else{
+                    togglebutton.data("status", "hidden").children("i").removeClass("fa-caret-down").addClass("fa-caret-right");
+                }
+            });
+            // open and close comments for one user
+            $("td").has(".comment-toggle").click(function(){
+                $(this).siblings().addBack().find(".full-comment").toggle();
+            });
+        });
+    </script>
+@endsection
```
Accessing this page generates more than 400 SQL queries, almost all of which are duplicate user queries. This slows down the page considerably. The number of duplicate queries should be 0 and the number of overall queries should be no more than 10. Of course, both of those numbers are soft limits and have some leeway, especially if there is a good reason or if the amount of work required to reduce the number is unreasonable. But 400 is too much.

The gig equivalent probably has the same issue.

nielsranosch commented on resources/views/layouts/app.blade.php:
```diff
@@ -141,6 +141,9 @@ className: 'help',
                                     <li>
                                         <a href="{{ route('rehearsals.listAttendances') }}">{{ trans('nav.attendance_last_rehearsal') }}</a>
                                     </li>
+                                    <li>
+                                        <a href="{{ route('rehearsals.listAttendances', ['id' => 'all']) }}">{{ trans('nav.attendance_rehearsals') }}</a>
+                                    </li>
```
I am not sure if this is intuitive for choir admins. However, I am willing to leave it as is, let them test it and make changes if necessary.

nielsranosch reviewed (changes requested):
Duplicate queries are bad, but a few are usually fine. 400 is definitely too much. If you need help reducing them, I habe time on the weekend.
